Fixes default classifier error with regression tests

chriski777 opened the pull request:
This PR addresses #476: 

- Adds module-specific fixture that copies the user's current classifier (if it exists) to a temporary directory, then copies over suite2p's default classifier to `.suite2p`. It then moves back the user's classifier to `.suite2p` during the teardown step of the test.

nickdelgrosso commented on tests/conftest.py:
```diff
@@ -16,3 +17,27 @@ def test_ops(tmpdir, data_dir):
     return ops
+
+
+@pytest.fixture(scope="module")
```
This fixture is very complex for so-specific a problem; it seems to me that it's a sign that the problem it is working around is a bug (or, to put it another way, a missing feature) in the source code.  A simpler approach would be helpful that way.

nickdelgrosso commented on README.md:
```diff
@@ -32,7 +32,7 @@ Install an [Anaconda](https://www.anaconda.com/download/) distribution of Python
 2. Open an anaconda prompt / command prompt with `conda` for **python 3** in the path
 3. Change directories to where the `environment.yml` is and run `conda env create -f environment.yml`
-4. Activate the environment with `conda activate suite2p'
+4. Activate the environment with `conda activate suite2p`
```
Good catch!

chriski777 commented:
> The solution presented here, while implemented nicely, is overcomplex in comparison to the problem being solved. A change in the source code (say, allowing one to specify they want to use the built-in default classifier, even if another default has been saved), would make the test suite simpler while providing a useful feature to the user.

Yeah, I agree. There is an approach that involves adding an ops parameter that contains a filepath to your desired classifier. I chose not to go with it because I didn't want to add another parameter to ops. But, seeing how complicated this fixture is, I'll try going with the approach above. Thanks for the feedback!

chriski777 commented:
Okay, so I managed to remove the fixture and replace it with simply feeding in `classfile=ops.get('classifier_path')` in `run_s2p.py`. This change shouldn't break with past version ops since `get` should just default to `None` if that key doesn't exist. 

I've also added a few refactoring changes to `classify.py`. 

-  Fixed issue with [mutable default parameter](https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/) `keys`
- replaced os import with pathlib

carsen-stringer commented on suite2p/classification/classify.py:
```diff
@@ -1,53 +1,50 @@
     """
+    if keys is None:
+        keys = []
+    keys = keys + ['npix_norm', 'compact', 'skew']
```
this is wrong, you might need to use fewer keys --- why can't we use my implementation of this?

chriski777 replied:
Yep, you're right. I didn't consider the option of having fewer keys. 

Having `keys`' default parameter as a mutable is dangerous as shown in this [link](https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/). For instance, the following code block fails: 

```
from suite2p import classification
import numpy as np
import copy
# stat with npix_norm, skew, compact
stat_file = np.load('/Users/chriski/Desktop/suite2p_ws/suite2p/plane0/stat.npy', allow_pickle=True)

# remove npix_norm
stat_no_npix = copy.deepcopy(stat_file)
for k in stat_no_npix:
    k.pop('npix_norm', None)

# remove skew from stat
stat_no_skew = copy.deepcopy(stat_file)
for ns in stat_no_skew:
    ns.pop('skew', None)

stat_no_compact = copy.deepcopy(stat_file)
for nc in stat_no_compact:
    nc.pop('compact', None)

classification.classify(save_path='/Users/chriski/Desktop/suite2p_ws', stat=stat_no_npix)
classification.classify(save_path='/Users/chriski/Desktop/suite2p_ws', stat=stat_no_skew)
classification.classify(save_path='/Users/chriski/Desktop/suite2p_ws', stat=stat_no_compact)

# just to check that stat_file has the default keys
print("stat_file has npix_norm value: {}".format(stat_file[0].get('npix_norm')))
print("stat_file has skew value: {}".format(stat_file[0].get('skew')))
print("stat_file has compact value: {}".format(stat_file[0].get('compact')))
classification.classify(save_path='/Users/chriski/Desktop/suite2p_ws', stat=stat_file)

```

The `remove` in classify.py ends up making the default `keys` list completely empty.

chriski777 replied:
Maybe it might be better to have default `keys` be a tuple and then convert it to a list within the classify func so that the keys could be removed safely??

nickdelgrosso commented:
This is definitely an improvement, great work @chriski777 !  I'm fine accepting it, but if you're interested we could I think it could be improved a little more, as this solution has the following complications:
  - there are now two different ways of getting to the built-in classifier npy file, and the way to specify it changes depending on whether a new default has been set, implying that the built-in npy loses its "default" status permanently as soon as a new default has been set.
  - The path to the npy file is now hard-coded into the test suite, so if it changes to a different location (for example, if we seperate the data from the python code, as may be good to do in the future), the tests will break, showing coupling of the code's organization with its behavior.
  - Users will have to find their suite2p installation folder in order to specify the path to the old defaullt if they want to use it again, which may be inconvenient for them, plus it means that we really can't change the location of the built-in npy, because some users will depend on it being where it is and named the way it is.

There are certainly many ways to solve this.  One idea may be to use a boolean ops key, "use_builtin_classifier", that defaults to False, and if True uses the built-in classifier.npy and ignores the user-specified default.  That way the test suite and users don't have to know the path in order to use it.

chriski777 commented:
Yeah, @nickdelgrosso, having a built-in classifier ops key seems like a useful feature. I've added that in the commits above.

nickdelgrosso commented:
thank you @chriski777 !   Looking at the classification package closer, this seems like a good opportunity to simplify the classify() function.  Let's pair on it today and merge afterward.

carsen-stringer commented:
can we reinstate the classify function please? all the code is now in run_s2p for determining paths but I liked it better the way you had it before

chriski777 commented:
Sure, would it be enough to just revert the last commit? That's the one where we removed classify() and replaced its calls

carsen-stringer commented:
I guess the last few? So you define the classfiles in there to make it easier for the user to run it

chriski777 commented:
Oh okay, the classfiles can be moved to classify. Could we keep the classfile-determining portions in run_s2p though? It seems that part is mostly relevant at the pipeline level. It might also be helpful for users to specify/know what classfile theyre using when they call classify().
